PythonVisualizations/AdvancedSorting.py

```diff
@@ -0,0 +1,570 @@
+                callEnviron.add(foundCircle)
+
+                # update the display
+                self.wait(0.3)
+
+                # remove the found circle
+                callEnviron.remove(foundCircle)
+                self.canvas.delete(foundCircle)
+
+                # Slide value rectangle up and off screen
+                items = (n.display_shape, n.display_val) if not self.changeSize else (n.display_shape,)
+                self.moveItemsOffCanvas(items, N, sleepTime=0.02)                
+
+                # Create an index for shifting the cells
+                kIndex = self.createIndex(i, 'k')
+                callEnviron |= set(kIndex)
+            
+                # Slide values from right to left to fill gap
+                for j in range(i+1, len(self.list)):
+                    self.assignElement(j, j - 1, callEnviron)
+                    self.fixCells() # retain the correct shape
+                    self.moveItemsBy(kIndex, (self.CELL_WIDTH, 0), sleepTime=0.01)
+                    self.wait(0.1)
+
+                # remove the last item in the list
+                n = self.list.pop()
+                # delete the associated display objects
+                self.canvas.delete(n.display_shape)
+                self.canvas.delete(n.display_val)
+                
+                # update window
+                self.wait(0.3)
+
+                self.cleanUp(callEnviron)
+                return True
+
+            # if not found, then move the index over one cell
+            self.moveItemsBy(indexDisplay, (self.CELL_WIDTH, 0), sleepTime=0.01)
+            if self.wait(0.1):
+                break
+
+        # Animation stops
+        self.cleanUp(callEnviron)
+        return None
+
+    
+    def new(self, val):
+        # if changeSize is equal to False
+        if self.changeSize:
+            # set CELL_WIDTH to be equal to CELL_SIZE
+            self.CELL_WIDTH = self.CELL_SIZE
+            # keep changeSize equal to False
+            self.changeSize = False
+        # if there is no room on the canvas  
+        if self.window.winfo_width() <= self.ARRAY_X0 + (val+1) * self.CELL_WIDTH:
+            # change the cell width
+            self.CELL_WIDTH = 20
+            # changeSize is set to True 
+            self.changeSize = True
+            # even when the cell width is smaller, return False if there is no room on the canvas 
+            if self.window.winfo_width() <= self.ARRAY_X0 + (val+1) * self.CELL_WIDTH:
+                return False
+            
+        callEnviron = self.createCallEnvironment()                
+        # Clear Array and reset size and list
+        self.canvas.delete("all")
+        self.size = val
+        self.list = []        
+        
+        for i in range(val):  # Draw new grid of cells
+            self.createArrayCell(i) 
+        
+        self.window.update()
+        self.cleanUp(callEnviron)
+        # return True upon success 
+        return True

```

Comment:
Let's move the New array operation to `SortingBase` since it is shared among many visualizations.  Let me propose simplifying it too.  It should just create an empty array of the right size and call `display()` to show it.  The size of the cells can be decided by the number of requested cells.
```
        self.size = val
        self.list = [None] * self.size
        canvasDimensions = self.widgetDimensions(self.canvas)
        self.changeSize = (self.size + 2) * self.CELL_SIZE > canvasDimensions[0]
        self.CELL_WIDTH = 20 if self.changeSize else self.CELL_SIZE
        self.display()
```

There should also be some maximum number of cells allowed, that can be customized by the subclasses of `SortingBase`.  Maybe a good default maximum would be 100 cells.  The `new()` method should verify the requested size is between 1 cell and the maximum and set the message if its outside that range.

Reply:
Fixed in commit 194409f